contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -2062,3 +2067,290 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
 	rewrite_object_refs(stmt);
 	return res;
 }
+
+Datum sp_babelfish_volatility(PG_FUNCTION_ARGS)
+{
+	int rc;
+	int i;
+	char *db_name = get_cur_db_name();
+	char *physical_schema_name = NULL;
+	char *function_signature = NULL;
+	char *query = NULL;
+	char *function_name = PG_ARGISNULL(0) ? NULL : TextDatumGetCString(PG_GETARG_TEXT_PP(0));
+	char *volatility = PG_ARGISNULL(1) ? NULL : TextDatumGetCString(PG_GETARG_TEXT_PP(1));
+	char *tsql_schema_name = NULL;
+	Oid function_id;
+	Oid user_id = GetUserId();
+
+	if(function_name != NULL)
+	{	
+		/* strip trailing whitespace */
+		i = strlen(function_name);
+		while (i > 0 && isspace((unsigned char) function_name[i - 1]))
+			function_name[--i] = '\0';
+
+		/* if function name is empty */
+		if(i == 0)
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					errmsg("function name is not valid")));
+		
+		/* length should be restricted to 4000 */
+		if (i > 4000)
+			ereport(ERROR,
+				(errcode(ERRCODE_STRING_DATA_LENGTH_MISMATCH),
+				errmsg("input value is too long for function name")));
+	}
+	if(volatility != NULL)
+	{
+		/* strip trailing whitespace */
+		i = strlen(volatility);
+		while (i > 0 && isspace((unsigned char) volatility[i - 1]))
+			volatility[--i] = '\0';
+		
+		/* if volatility is empty */
+		if(i==0)
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					errmsg("volatility is not valid")));
+
+		/* its length is greater than 9 (len of immutable) */
+		if (i > 9)
+			ereport(ERROR,
+				(errcode(ERRCODE_STRING_DATA_LENGTH_MISMATCH),
+				errmsg("input value is too long for volatility")));
+	}
+	if(function_name == NULL && volatility != NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+				errmsg("function name cannot be NULL")));
+
+	if(function_name != NULL)
+	{	
+		List *function_name_list;
+		FuncCandidateList candidates = NULL;
+		char *full_function_name = NULL;
+		char *logical_schema_name = NULL;
+		char **splited_object_name;
+
+		/* get physical schema name */
+		splited_object_name = split_object_name(function_name);
+
+		if(strcmp(splited_object_name[0], "") || strcmp(splited_object_name[1], ""))
+			ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						errmsg("function \"%s\" is not a valid two part name", function_name)));
+		
+		pfree(function_name);
+		logical_schema_name = splited_object_name[2];
+		function_name = splited_object_name[3];
+		pfree(splited_object_name);
+		
+		/* downcase identifier */
+		if(pltsql_case_insensitive_identifiers)
+		{
+			logical_schema_name = downcase_identifier(logical_schema_name, strlen(logical_schema_name), false, false);
```

Comment:
downcase_identifier creates new allocation of memory. Better to use splited_object_name[2]/[3] and then free splited_object_name and logical_schema_name separately